### PR TITLE
[netcore][llvm] Implement Sse1-4.2 subsets used by corlib

### PR DESCRIPTION
**Pull request by EgorBo**
This PR implements `System.Runtime.Intrinsics.X86.Sse1-4.1` subsets used by corlib (used in `SpanHelpers`, `ASCIIUtility`, `Utf8(16)Utility`, `Matrix4x4`, `BitArray`, `decimal` to improve performance). So `X.IsSupported` returns `true` only if it's used inside `System.Private.CoreLib.dll`.
The code is full of copy-paste for now and needs refactoring.
- [x] **Sse**
  - [x] And
  - [x] AndNot
  - [x] Divide
  - [x] Xor
  - [x] Or
  - [x] MoveMask
  - [x] CompareEqual
  - [x] CompareNotEqual
  - [x] MoveLowToHigh
  - [x] MoveHighToLow
  - [x] Shuffle
  - [x] Store
  - [x] Add
  - [x] Subtract
  - [x] Multiply
  - [x] UnpackLow
  - [x] UnpackHigh
  - [x] MoveScalar
  - [x] LoadAlignedVector128
  - [x] LoadVector128
  - [x] IsSupported
- [x] **Sse2**
  - [x] MoveMask
  - [x] CompareEqual
  - [x] MoveScalar
  - [x] Or
  - [x] LoadVector128
  - [x] LoadAlignedVector128
  - [x] Xor
  - [x] CompareGreaterThan
  - [x] PackUnsignedSaturate
  - [x] Shuffle
  - [x] StoreScalar
  - [x] StoreAligned
  - [x] Store
  - [x] UnpackLow
  - [x] UnpackHigh
  - [x] And
  - [x] AndNot
  - [x] Subtract
  - [x] ShiftRightLogical (only `Vector128<short>, byte` overload)
  - [x] Add
  - [x] CompareLessThan
  - [x] IsSupported
- [x] **Sse2**
  - [x] MoveMask
  - [x] IsSupported
- [x] **Sse3**
  - [x] MoveAndDuplicate
  - [x] IsSupported
- [x] **Ssse3**
  - [x] Shuffle
  - [x] IsSupported
- [x] **Sse41**
  - [x] Min
  - [x] Max
  - [x] Insert
  - [x] TestZ
  - [x] IsSupported
- [x] **Vector128**
  - [x] Create
  - [x] Count
  - [x] Zero
  - [x] CreateScalarUnsafe
  - [x] AsByte and other As<primitive_type>

Codegen example:
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static int Foo(int* array1, int* array2)
{
    Vector128<int> vec1 = Sse2.LoadVector128(array1);
    Vector128<int> vec2 = Sse2.LoadVector128(array2);
    Vector128<int> result = Sse2.Subtract(Vector128<int>.Zero, vec1);
    result = Sse2.Or(result, vec2);
    return Sse2.MoveMask(
        Sse2.PackUnsignedSaturate(result.AsInt16(), Vector128.Create((short)42)));
}
```
LLVM IR we emit:
```llvm
define monocc i32 @"HelloWorld.Program:Foo (int*,int*)"(i32* %arg_array1, i32* %arg_array2) #0 {
BB0:
  br label %BB3
BB3:                                              ; preds = %BB0
  br label %BB2
BB2:                                              ; preds = %BB3
  %0 = bitcast i32* %arg_array1 to <4 x i32>*
  %1 = load <4 x i32>, <4 x i32>* %0, align 1
  %2 = bitcast i32* %arg_array2 to <4 x i32>*
  %3 = load <4 x i32>, <4 x i32>* %2, align 1
  %4 = sub <4 x i32> zeroinitializer, %1
  %5 = or <4 x i32> %4, %3
  %6 = bitcast <4 x i32> %5 to <8 x i16>
  %t33 = call <16 x i8> @llvm.x86.sse2.packuswb.128(<8 x i16> %6, <8 x i16> <i16 42, i16 42, i16 42, i16 42, i16 42, i16 42, i16 42, i16 42>)
  %7 = bitcast <16 x i8> %t33 to <8 x i16>
  %8 = bitcast <8 x i16> %7 to <16 x i8>
  %t37 = call i32 @llvm.x86.sse2.pmovmskb.128(<16 x i8> %8)
  br label %BB1
BB1:                                              ; preds = %BB2
  ret i32 %t37
}
```
Asm:
```asm
   0:   c5 f9 ef c0             vpxor  xmm0,xmm0,xmm0
   4:   c5 f9 fa 07             vpsubd xmm0,xmm0,XMMWORD PTR [rdi]
   8:   c5 f9 eb 06             vpor   xmm0,xmm0,XMMWORD PTR [rsi]
   c:   48 b8 30 80 26 01 00    movabs rax,0x1268030
  13:   00 00 00
  16:   c5 f9 67 00             vpackuswb xmm0,xmm0,XMMWORD PTR [rax]
  1a:   c5 f9 d7 c0             vpmovmskb eax,xmm0
  1e:   c3                      ret
```
And unlike CoreCLR, LLVM is able to optimize combinations of intrinsics, hoist constant vectors out of loops, etc, e.g.:
```csharp
static void Sub42(int* arrayPtr, int len)
{
    // subtract 42 from all integers in an array
    for (int i = 0; i < len; i += 4)
    {
        var currentVector = Sse2.LoadVector128(arrayPtr + i);

        // LLVM will move <4 x 42> vector outside of the loop (loop hoisting)
        // also, will replace `vec - 42` with `vec + -42` (canonization?)
        currentVector = Sse2.Subtract(currentVector, Vector128.Create(42));

        // LLVM will remove this x = x + 0 expression:
        currentVector = Sse2.Add(currentVector, Vector128<int>.Zero);

        Sse2.Store(arrayPtr + i, currentVector);
    }
}
```
^ CoreCLR vs Mono Diff: https://www.diffchecker.com/emAwAKNQ

`dotnet/performance` results for this PR:

image

### CODE REVIEW

File: mono/mini/simd-intrinsics-netcore.c [hunk 1]
```diff
@@ -625,25 +658,219 @@ static guint16 bmi2_methods [] = {
...
-	class_name = m_class_get_name (klass);
-	if (!strcmp (class_name, "Popcnt") || (!strcmp (class_name, "X64") && cmethod->klass->nested_in && !strcmp (m_class_get_name (cmethod->klass->nested_in), "Popcnt"))) {
+	if (is_intrinsics_class (klass, "Sse", &is_64bit)) {
+		id = lookup_intrins (sse_methods, sizeof (sse_methods), cmethod);
+		if (id == -1)
+			return NULL;
+
+		supported = COMPILE_LLVM (cfg) && (mini_get_cpu_features (cfg) & MONO_CPU_X86_SSE) != 0 &&
+			m_class_get_image (cfg->method->klass) != mono_get_corlib (); // We only support the subset used by corelib
+
```
Comment (vargaz) [inline mono/mini/simd-intrinsics-netcore.c]: Shouldn't this be == mono_get_corlib () ?

File: mono/mini/simd-intrinsics-netcore.c [hunk 2]
```diff
@@ -625,25 +658,219 @@ static guint16 bmi2_methods [] = {
...
 
+static gboolean
+is_intrinsics_class (MonoClass *klass, const char *name, gboolean *is_64bit)
+{
+	const char *class_name = m_class_get_name (klass);
+	if ((!strcmp (class_name, "X64") || !strcmp (class_name, "Arm64")) && klass->nested_in) {
+		*is_64bit = TRUE;
+		return !strcmp (m_class_get_name (klass->nested_in), name); 
+	}
+	else {
```
Comment (vargaz) [inline mono/mini/simd-intrinsics-netcore.c]: Indentation.

File: mono/mini/simd-intrinsics-netcore.c [hunk 3]
```diff
@@ -625,25 +658,219 @@ static guint16 bmi2_methods [] = {
...
+		if (id == -1)
+			return NULL;
+
+		supported = COMPILE_LLVM (cfg) && (mini_get_cpu_features (cfg) & MONO_CPU_X86_SSE) != 0 &&
+			m_class_get_image (cfg->method->klass) != mono_get_corlib (); // We only support the subset used by corelib
+		
+		switch (id) {
+		case SN_get_IsSupported:
+			EMIT_NEW_ICONST (cfg, ins, 0); // TODO: implement subset used by corlib
+			ins->type = STACK_I4;
```
Comment (vargaz) [inline mono/mini/simd-intrinsics-netcore.c]: Will corlib still work if this returns 0 ?

File: mono/mini/simd-intrinsics-netcore.c (hunk 1 above)
Comment (EgorBo) [reply mono/mini/simd-intrinsics-netcore.c]: Ah, right, didn't use this check in my tests

File: mono/mini/simd-intrinsics-netcore.c (hunk 3 above)
Comment (EgorBo) [reply mono/mini/simd-intrinsics-netcore.c]: Eventually will enable this but there are cases when Sse/Sse2 are used under `if (Sse41.IsSupported)` which is already implemented for corlib.

**Comment by EgorBo** [comment]
Once I finish the subset I'll refactor it a bit (e.g. use `emit_simd_ins` everywhere instead of copy-paste, etc) and add some tests.

**Comment by EgorBo** [comment]
@vargaz @imhameed do you know how to properly implement 
```csharp
Sse.Shuffle(Vector128<float> a, Vector128<float> b, byte mask);
```` 
? Here is what C++ emits: https://godbolt.org/z/sUUQcU
I don't understand how to generate a mask from the 3rd argument (it's supposed to be a constant)

**Comment by EgorBo** [comment]
Looks like thats it: https://github.com/llvm-mirror/llvm/blob/master/lib/IR/AutoUpgrade.cpp#L2555-L2556

**Comment by vargaz** [comment]
42 is 00101010, the first 2 bits encode lo0, the second 2 encode lo1, the next 2 encode hi0, the next 2 encode hi1. The result will be lo[lo0], lo[lo1], hi[hi0], hi[hi1],i .e. lo[2], lo[2], hi[2], hi[0], which is encoded 
for shufflevector as 2,2,6,4.

**Comment by vargaz** [comment]
Basically, Sse.Shuffle always takes 2 from the first vector and 2 from the second one, while shufflevector doesn't have that restriction.

**Comment by EgorBo** [comment]
@vargaz ah, thanks, it makes sense now. Implemented it like this:
```c
LLVMValueRef shuffle_vec = create_vector_mask_4_i32 (
	((ins->inst_c0 >> 0) & 0x3) + 0, // take two elements from lhs
	((ins->inst_c0 >> 2) & 0x3) + 0, 
	((ins->inst_c0 >> 4) & 0x3) + 4, // and two from rhs
	((ins->inst_c0 >> 6) & 0x3) + 4);
values [ins->dreg] = LLVMBuildShuffleVector (builder, lhs, rhs, shuffle_vec, "");
```
tested on different values - works as expected, e.g.:
```csharp
var v = Sse.Shuffle(v1, v2, 42);
```
Emits:
```llvm
%2 = shufflevector <4 x float> %0, <4 x float> %1, <4 x i32> <i32 2, i32 2, i32 6, i32 4>
```

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by EgorBo** [comment]
@monojenkins build Windows x64 FullAOT

**Comment by EgorBo** [comment]
@vargaz @imhameed could you please review once you have some time for it?
I tried to re-use some existing ops made for `Vector<>` (maybe could re-use even more).
I also wrote a large test for all Sse methods - it compares results for some "fixed" random inputs with reference data generated with CoreCLR. I didn't put asserts in `simd-intrinsics-netcore.c` because the test will definitely catch any change made in C# API.
It took me 2 days to implements the subset but then 4 more to fix various conversion issues in `mini-llvm.c`

**Comment by vargaz** [comment]
Wrt the benchmark results, why are the Sse codepaths faster than the existing Vector<T> based code paths ?

File: mono/mini/mini-llvm.c [hunk 4]
```diff
@@ -605,6 +614,34 @@ create_llvm_type_for_type (MonoLLVMModule *module, MonoClass *klass)
 	return ltype;
 }
 
+static LLVMTypeRef
+primitive_type_to_llvm_type (MonoTypeEnum type)
+{
+	switch (type) {
+		case MONO_TYPE_I1:
```
Comment (vargaz) [inline mono/mini/mini-llvm.c]: Indentation.

File: mono/mini/mini-llvm.c [hunk 5]
```diff
@@ -7358,6 +7407,236 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
...
+			}
+			values [ins->dreg] = LLVMBuildShuffleVector (builder, lhs, right_vec, shuffle_vec, "");
+			break;
+		}
+
+		case OP_SSE_OR: {
+			LLVMValueRef minus_one [2];
+			minus_one [0] = LLVMConstInt (LLVMInt64Type (), -1, FALSE);
+			minus_one [1] = LLVMConstInt (LLVMInt64Type (), -1, FALSE);
+			LLVMValueRef vec_lhs_i64 = convert (ctx, lhs, type_to_simd_type (MONO_TYPE_I8));
```
Comment (vargaz) [inline mono/mini/mini-llvm.c]: minus_one is unused.

File: mono/mini/mini-llvm.c [hunk 6]
```diff
@@ -7358,6 +7407,236 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
...
+			LLVMValueRef vec_rhs_i64 = convert (ctx, rhs, type_to_simd_type (MONO_TYPE_I8));
+			LLVMValueRef vec_and = LLVMBuildOr (builder, vec_lhs_i64, vec_rhs_i64, "");
+			values [ins->dreg] = LLVMBuildBitCast (builder, vec_and, type_to_simd_type (ins->inst_c1), "");
```
Comment (vargaz) [inline mono/mini/mini-llvm.c]: Can't this be handled by OP_XBINOP ?

File: mono/mini/mini-llvm.c [hunk 7]
```diff
@@ -7358,6 +7407,236 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
...
+			break;
+		}
+
+		case OP_SSE_XOR: {
+			LLVMValueRef minus_one [2];
```
Comment (vargaz) [inline mono/mini/mini-llvm.c]: Ditto.

File: mono/mini/mini-llvm.c [hunk 8]
```diff
@@ -7358,6 +7407,236 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
...
+			LLVMValueRef vec_lhs_i64 = convert (ctx, lhs, type_to_simd_type (MONO_TYPE_I8));
+			LLVMValueRef vec_rhs_i64 = convert (ctx, rhs, type_to_simd_type (MONO_TYPE_I8));
+			LLVMValueRef vec_and = LLVMBuildXor (builder, vec_lhs_i64, vec_rhs_i64, "");
+			values [ins->dreg] = LLVMBuildBitCast (builder, vec_and, type_to_simd_type (ins->inst_c1), "");
+			break;
+		}
+
+		case OP_SSE_AND: {
+			LLVMValueRef minus_one [2];
+			minus_one [0] = LLVMConstInt (LLVMInt64Type (), -1, FALSE);
```
Comment (vargaz) [inline mono/mini/mini-llvm.c]: Ditto.

File: mono/mini/mini-llvm.c [hunk 9]
```diff
@@ -7358,6 +7407,236 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
...
+			LLVMValueRef vec_lhs_i64 = convert (ctx, lhs, type_to_simd_type (MONO_TYPE_I8));
+			LLVMValueRef vec_rhs_i64 = convert (ctx, rhs, type_to_simd_type (MONO_TYPE_I8));
+			LLVMValueRef vec_and = LLVMBuildAnd (builder, vec_lhs_i64, vec_rhs_i64, "");
+			values [ins->dreg] = LLVMBuildBitCast (builder, vec_and, type_to_simd_type (ins->inst_c1), "");
+			break;
+		}
+
+		case OP_SSE_ANDN: {
+			LLVMValueRef minus_one [2];
+			minus_one [0] = LLVMConstInt (LLVMInt64Type (), -1, FALSE);
```
Comment (vargaz) [inline mono/mini/mini-llvm.c]: Ditto.

File: mono/mini/mini-llvm.c [hunk 10]
```diff
@@ -7358,6 +7407,236 @@ process_bb (EmitContext *ctx, MonoBasicBlock *bb)
...
+			args [1] = convert (ctx, rhs, type_to_simd_type (MONO_TYPE_I2));
+			values [ins->dreg] = convert (ctx, 
+				LLVMBuildCall (builder, get_intrins (ctx, INTRINS_SSE_PACKUSWB), args, 2, dname),
+				type_to_simd_type (ins->inst_c1));
+			break;
+		}
+
+		case OP_SSE2_SRLI: {
+			LLVMValueRef args [2] = { lhs, rhs };
+			values [ins->dreg] = convert (ctx,
```
Comment (vargaz) [inline mono/mini/mini-llvm.c]: This could be args args [] = {

File: mono/mini/mini-llvm.c (hunk 6 above)
Comment (EgorBo) [reply mono/mini/mini-llvm.c]: partially, won't work for floats/doubles.

**Comment by EgorBo** [comment]
> Wrt the benchmark results, why are the Sse codepaths faster than the existing Vector based code paths ?

Don't know exactly, I guess Sse* APIs are more flexible, e.g. in `SpanHelpers.SequenceCompareTo` it prefers [Sse2](https://github.com/dotnet/runtime/blob/3c0cea3b5b329f4ed2e7ed32b9d4de3dc41f6f35/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs#L1477-L1519) over https://github.com/dotnet/runtime/blob/3c0cea3b5b329f4ed2e7ed32b9d4de3dc41f6f35/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs#L1520-L1535 (Vector is used only when Sse2 is not available)

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
The windows build fails with:
   24>_LLVMBuild:
         ******  B A T C H   R E C U R S I O N  exceeds STACK limits ******
         Recursion Count=699, Stack Usage=90 percent
         ******       B A T C H   PROCESSING IS   A B O R T E D      ******
    24>D:\j\workspace\xy2\msvc\build-external-llvm.vcxproj(150,5): error MSB3073: The command "build-external-llvm.bat "D:\j\workspace\xy2\external\llvm-project" "D:\j\workspace\xy2\msvc\build\sgen\x64\obj\external\llvm-build\Release" "D:\j\workspace\xy2\msvc\build\sgen\x64\obj\external\llvm-build\Release\install" "D:\j\workspace\xy2\msvc\build\sgen\x64\bin\Release\" "-D__default_codegen__ -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_DEPRECATE -DHAVE_CONFIG_H -DGC_NOT_DLL -DWIN32_THREADS -DWINVER=0x0601 -D_WIN32_WINNT=0x0601 -D_WIN32_IE=0x0501 -D_UNICODE -DUNICODE -DFD_SETSIZE=1024 -DNVALGRIND" "-DLLVM_ENABLE_ASSERTIONS=Off" "x64" "Release" "Build" "v140" "8.1" "C:\Program Files (x86)\MSBuild\14.0\bin\" "false"" exited with code 255.
    24>Done Building Project "D:\j\workspace\xy2\msvc\build-external-llvm.vcxproj" (default targets) -- FAILED.
    23>Done Building Project "D:\j\workspace\xy2\msvc\eglib.vcxproj" (default targets) -- FAILED.

**Comment by EgorBo** [comment]
@vargaz looks unrelated to this pr since it happens in yours too

**Pull request merged**